server/gokbg3/grails-app/services/org/gokb/TippService.groovy

```diff
@@ -21,18 +27,110 @@ class TippService {
   def reviewRequestService
   def autoTimestampEventListener
 
+  public def updateCoverage(tipp, reqBody) {
+    def cov_list = reqBody.coverageStatements ?: reqBody.coverage
+    def missing = tipp.coverageStatements.collect { it.id }
+    def changed = false
+
+    cov_list?.each { c ->
+      def parsedStart = GOKbTextUtils.completeDateString(c.startDate)
+      def parsedEnd = GOKbTextUtils.completeDateString(c.endDate, false)
+
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'startVolume', c.startVolume)
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'startIssue', c.startIssue)
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'endVolume', c.endVolume)
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'endIssue', c.endIssue)
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'embargo', c.embargo)
+      changed |= com.k_int.ClassUtils.setStringIfDifferent(tipp, 'coverageNote', c.coverageNote)
+      changed |= com.k_int.ClassUtils.setDateIfPresent(parsedStart, tipp, 'startDate')
+      changed |= com.k_int.ClassUtils.setDateIfPresent(parsedEnd, tipp, 'endDate')
+      changed |= com.k_int.ClassUtils.setRefdataIfPresent(c.coverageDepth, tipp, 'coverageDepth', 'TitleInstancePackagePlatform.CoverageDepth')
+
+      def cs_match = false
+      def startAsDate = (parsedStart ? Date.from(parsedStart.atZone(ZoneId.systemDefault()).toInstant()) : null)
+      def endAsDate = (parsedEnd ? Date.from(parsedEnd.atZone(ZoneId.systemDefault()).toInstant()) : null)
+
+      tipp.coverageStatements?.each { tcs ->
+
+        if (!cs_match && (
+          (c.id && tcs.id == c.id) ||
+            (tcs.startVolume && tcs.startVolume == c.startVolume) ||
+            (tcs.startDate && tcs.startDate == startAsDate) ||
+            (!cs_match && !tcs.startVolume && !tcs.startDate && !tcs.endVolume && !tcs.endDate))
+        ) {
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'startIssue', c.startIssue)
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'startVolume', c.startVolume)
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'endVolume', c.endVolume)
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'endIssue', c.endIssue)
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'embargo', c.embargo)
+          changed |= com.k_int.ClassUtils.setStringIfDifferent(tcs, 'coverageNote', c.coverageNote)
+          changed |= com.k_int.ClassUtils.setDateIfPresent(parsedStart, tcs, 'startDate')
+          changed |= com.k_int.ClassUtils.setDateIfPresent(parsedEnd, tcs, 'endDate')
```

Comment:
We want to use `updateDateField` instead, since we want to be able remove existing dates as well.

Reply:
done